tests/prefetch_related/tests.py

```diff
@@ -223,6 +224,27 @@ def test_invalid_final_lookup(self):
         self.assertIn('prefetch_related', str(cm.exception))
         self.assertIn("name", str(cm.exception))
 
+    def test_m2m_then_reverse_fk_object_ids(self):
+        with CaptureQueriesContext(connection) as queries:
+            list(Book.objects.prefetch_related('authors__addresses'))
+
+        sql = queries[-1]['sql']
+        self.assertEqual(sql.count(self.author1.name), 1)
+
+    def test_m2m_then_m2m_object_ids(self):
+        with CaptureQueriesContext(connection) as queries:
+            list(Book.objects.prefetch_related('authors__favorite_authors'))
+
+        sql = queries[-1]['sql']
+        self.assertEqual(sql.count(self.author1.name), 1)
+
+    def test_m2m_then_reverse_one_to_one_object_ids(self):
+        with CaptureQueriesContext(connection) as queries:
+            list(Book.objects.prefetch_related('authors__authorwithage'))
+
+        sql = queries[-1]['sql']
+        self.assertEqual(sql.count(str(self.author1.id)), 1, sql)
```

Comment:
I'm missing a case here for [`ForwardManyToOneDescriptor`](https://github.com/django/django/blob/c3531d2f208c5a039a190ec36a7349f627f13f01/django/db/models/fields/related_descriptors.py#L114-L143) which already [partially fixes](https://github.com/django/django/blob/c3531d2f208c5a039a190ec36a7349f627f13f01/django/db/models/fields/related_descriptors.py#L131) this issue.